source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+
+  std::vector<int> off_set(n_mpi_processes, 0);
+  int              total_size = 0;
+  for (unsigned int i = 0; i < n_mpi_processes; ++i)
+    {
+      off_set[i] = total_size;
+      total_size += recv_counts[i];
+    }
+
+  // Build full filename (same for all ranks)
+  std::string full_filename = folder + "/" + file_name + ".xyz";
+
+  // Open MPI file
+  MPI_File fh;
+  MPI_File_open(mpi_communicator,
+                full_filename.c_str(),
+                MPI_MODE_CREATE | MPI_MODE_WRONLY,
+                MPI_INFO_NULL,
+                &fh);
+
+  // Each rank writes at its own offset
+  MPI_File_write_at(fh,
+                    off_set[this_mpi_process],
+                    local_str.data(),
+                    (int)local_size,
+                    MPI_CHAR,
+                    MPI_STATUS_IGNORE);
+
+  MPI_File_close(&fh);
+}
+
+
+template <int dim>
+void
+RayTracingSolver<dim>::finish_simulation()
+{
+  // Timer output
+  if (parameters.timer.type == Parameters::Timer::Type::end)
+    this->computing_timer.print_summary();
+
+  // Testing
+  if (parameters.test.enabled)
+    {
+      // This needs to be coded later.
+    }
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::solve()
+{
+  // Set up the parameters
+  setup_parameters();
+
+  // Reading mesh
+  read_mesh(parameters.mesh,
+            action_manager->check_restart_simulation(),
+            pcout,
+            triangulation,
+            dem_parameters.boundary_conditions);
+
+  // Set particle insertion object
+  particle_insertion_object = set_particle_insertion_type();
+
+  // Insert photon and particles
+  insert_particles_and_photons();
+
+  // Particles won't move, thus we can sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  // Every intersection point from the beginning of the simulation.
+  std::vector<Point<dim>> total_intersection_points;
+  total_intersection_points.reserve(photon_handler.n_global_particles());
+
+  // Map :
+  // <photon_iterator, (distance, intersection point<dim>, iterator to remove)>
+  ankerl::unordered_dense::map<
+    types::particle_index,
+    std::tuple<double, Point<dim>, Particles::ParticleIterator<dim>>>
+    photon_intersection_points_map;
+
+  photon_intersection_points_map.reserve(
+    photon_handler.n_locally_owned_particles());
+
+  // Update cell neighbors
+  find_cell_neighbors<dim, true>(triangulation,
+                                 cells_local_neighbor_list,
+                                 cells_ghost_neighbor_list);
+
+  // Particle don't move, thus we sort them at the end of a load balance.
+  // Photons are being sorted every pseudo time step, thus no need to sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  while (photon_handler.n_global_particles() != 0)
```

Comment:
What I would do is I would add a SimulationControlRayTracing and at the end of each time-step I would provide the SimulationControlRayTracing with the number of photons remaining.